src/context.c

```diff
@@ -290,6 +291,171 @@ ly_ctx_new(const char *search_dir, uint16_t options, struct ly_ctx **new_ctx)
     return rc;
 }
 
+static LY_ERR
+ly_ctx_new_yl_legacy(struct ly_ctx **ctx, struct lyd_node *yltree)
+{
+    struct lyd_node *module, *node;
+    struct ly_set *set;
+    const char **feature_arr;
+    LY_ERR ret;
+
+    LY_CHECK_RET(ret = lyd_find_xpath(yltree, "/ietf-yang-library:yang-library/modules-state/module", &set), ret);
+
+    /* process the data tree */
+    for (uint32_t i = 0; i < set->count; ++i) {
+        module = set->dnodes[i];
+
+        /* initiate */
+        const char *name = NULL, *revision = NULL;
+        struct ly_set features = {0};
+        ly_bool imported = 0;
+
+        LY_LIST_FOR(*lyd_node_child_p(module), node) {
+            if (!strcmp(node->schema->name, "name")) {
+                name = ((struct lyd_node_term *)node)->value.canonical;
+            } else if (!strcmp(node->schema->name, "revision")) {
+                revision = ((struct lyd_node_term *)node)->value.canonical;
+            } else if (!strcmp(node->schema->name, "feature")) {
+                LY_CHECK_ERR_RET(ret = ly_set_add(&features, node, 0, NULL),
+                        ly_set_free(&features, free); ly_set_free(set, free), ret);
+            } else if (!strcmp(node->schema->name, "conformance-type") &&
+                    ((struct lyd_node_any *)node)->meta->value.enum_item->value) {
+                /* imported module - skip it, it will be loaded as a side effect
+                 * of loading another module */
+                imported = 1;
+                break;
+            }
+        }
+
+        if (imported) {
+            continue;
+        }
+
+        feature_arr = malloc(sizeof(char *) * (features.count + 1));
```

Comment:
Avoid `sizeof(type)` whenever possible, always prefer using a variable directly (if you change its type later, the size will change automatically). So write `sizeof *feature_arr` instead.